Sources/SwiftProtobuf/_MessageStorage.swift

```diff
@@ -719,6 +719,141 @@ extension _MessageStorage {
+                        _MessageLayout.SubmessageToken(index: field.submessageIndex),
+                        field,
+                        self,
+                        other
+                    )
+                case .int32, .sfixed32, .sint32:
+                    equalSoFar = isField(field, equalToSameFieldIn: other, type: [Int32].self)
+                case .int64, .sfixed64, .sint64:
+                    equalSoFar = isField(field, equalToSameFieldIn: other, type: [Int64].self)
+                case .string:
+                    equalSoFar = isField(field, equalToSameFieldIn: other, type: [String].self)
+                default:
+                    preconditionFailure("Unreachable")
+                }
+
+            case .scalar:
+                switch field.rawFieldType {
+                case .bytes:
+                    if field.offset < firstNontrivialStorageOffset {
+                        firstNontrivialStorageOffset = field.offset
+                    }
+                    equalSoFar = isField(field, equalToSameFieldIn: other, type: Data.self)
+
+                case .group, .message:
+                    if field.offset < firstNontrivialStorageOffset {
+                        firstNontrivialStorageOffset = field.offset
+                    }
+                    equalSoFar = layout.areSubmessagesEqual(
+                        _MessageLayout.SubmessageToken(index: field.submessageIndex),
+                        field,
+                        self,
+                        other
+                    )
+
+                case .string:
+                    if field.offset < firstNontrivialStorageOffset {
+                        firstNontrivialStorageOffset = field.offset
+                    }
+                    equalSoFar = isField(field, equalToSameFieldIn: other, type: String.self)
+
+                default:
+                    // Do nothing. Trivial fields will be bitwise-compared as a block below.
+                    break
+                }
+
+            default:
+                preconditionFailure("Unreachable")
+            }
+
+            guard equalSoFar else {
+                return false
+            }
+        }
+
+        // Compare all of the trivial values (including has-bits) in bitwise fashion.
+        if firstNontrivialStorageOffset != 0 {
+            return memcmp(buffer.baseAddress!, other.buffer.baseAddress!, firstNontrivialStorageOffset) == 0
+        }
+        return true
+    }
+
+    /// Returns whether the given field in the receiver is equal to the same field in the other
+    /// storage, given the expected type of that field.
+    public func isField<T: Equatable>(
+        _ field: FieldLayout,
+        equalToSameFieldIn other: _MessageStorage,
+        type: T.Type
+    ) -> Bool {
+        let isSelfPresent = isPresent(field)
+        let isOtherPresent = other.isPresent(field)
+        if !isSelfPresent && !isOtherPresent {
```

Comment:
This might be too cute/magic (assuming I got it right), but it streamlines the both set cases.

```
// If they both aren't set, then they are only equal if they both aren't set.
guard isSelfPresent && isOtherPresent else {
  return isSelfPresent == isOtherPresent
}
```

Reply:
I think that's reasonable, along with a comment to make it seem less magical.